Wrap the Banner

jordankoschei-okta opened the pull request:
In hopes of helping accelerate our component-wrapping work, I tried my hand at wrapping a simple component. This PR is my first code for Okta, and my first attempt at Typescript, so please review my style and best practices.

**What's included:**
- I wrapped the banner-type alert in a new `<Banner>` component
- Added logic for showing the action link as necessary
- Swapped the old Storybook stories to use `<Banner>`

**What I'm not sure about:**
- Are the types for each prop accurate?
- The link logic depends on whether the component has the `linkUrl` prop set, so if the component has a `linkText` but no `linkUrl`, it won't display. This feels clunky — is there a better way of doing this?
- Do we need to allow the user to add arbitrary props to the `<Link>`? I don't want to give consumers the option to change the link variant from `monochrome`, but it would be nice if users could add `target="blank"` if needed.

edburyenegren-okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,41 @@
+  severity?: AlertColor;
+  role?: string;
+  children: any;
+  onClose?: any;
```
Rather than expose `onClose` directly, should we include the JS for dismissal and change this to a boolean called `isDismissible`? That way, we'll know the behavior is consistent.

edburyenegren-okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,41 @@
+export interface BannerProps {
+  severity?: AlertColor;
+  role?: string;
+  children: any;
```
Now that we're separating out the Link (good call), should we change this to `string`?

edburyenegren-okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,41 @@
+
+export interface BannerProps {
+  severity?: AlertColor;
+  role?: string;
```
I'm not sure what the syntax looks like, but we should probably limit this to `undefined` and status`.

I think `role?: "status" | undefined;`?

jordankoschei-okta commented:
At @edburyenegren-okta's suggestion, I refactored the component to have a boolean `isDismissable` prop rather than an `onClose` prop that accepts a function. This ensures that we have full control over the close behavior, rather than relying on consumers to include the proper function.

I also added explanatory comments to the type definitions, following @KevinGhadyani-Okta's pattern from `<TextField>`.

jordankoschei-okta commented:
Fixed merge conflicts.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,79 @@
+   * ("status" for something that dynamically updates, null for something
+   * unchanging)
+   */
+  role?: "status" | undefined;
```
`role` is an HTML term, so I wouldn't recommend using it for something other than accessibility roles.

edburyenegren-okta replied:
MUI uses this prop to set the HTML attribute under the hood.

> role:	The ARIA role attribute of the element.

(from https://mui.com/material-ui/api/alert/)

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,79 @@
+}: BannerProps) => {
+  const [isVisible, setIsVisible] = useState(true);
+
+  const handleClose = () => setIsVisible(false);
```
```suggestion
  const handleClose = useCallback(() => setIsVisible(false), []);
```

Import `useCallback` from `'react'`.

KevinGhadyani-Okta replied:
We should also rename this to something like `closeAlert` instead of `handleClose` because `handleClose` doesn't tell me anything about what the function does.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  const handleClose = () => setIsVisible(false);
+
+  return (
+    <React.Fragment>
+      {isVisible && (
+        <Alert
+          role={role}
+          variant="banner"
+          severity={severity}
+          onClose={isDismissable ? handleClose : undefined}
+        >
```
I really don't think we should be handling the Alert visibility in the alert itself, it should be handled by the renderer.

This is a cross-pollination of roles. Typically, consumers should handle the rendering of components, not the components themselves. You can create a hook with the logic necessary for consuming it.

There are better ways of doing this with a children-first pattern I've talked about, but that's not a pattern we use at Okta, and it would require a more diverse set of tools rather than per-component logic.

At this point, I'd recommend making visibility something handled outside of `Banner`. You can provide a hook outside of it. You'll also want to add a callback for when a dismissal occurred so the logic outside can dismiss the banner.

Either you or someone else can add the necessary hook with visibility logic. It could be `useBanner`, but it'd best be something generic like `useVisibility`. I've written a hook like this already in my open-source library of Children-First Components: https://github.com/Sawtaytoes/children-first-components/blob/master/src/useVisibility.ts.

We don't have to take it verbatim, but we should discuss possible solutions rather than building hide-show logic into every Odyssey component.

Another option is following MUI's take and doing it however they do it, but there's nothing stopping MUI from bad-design patterns that, because of copy-paste-modify, will be replicated throughout our apps at Okta.

---

Lastly, we need to change this hide-show functionality because you have no way of telling the consumer when to remove the banner or how to show it again.

Right now, it gets rendered when an error occurs and then hides itself, but the consumer still has it rendered. If another error occurs, it won't show because it's already rendered, and the consumer has no way of passing a prop saying "show again".

jordankoschei-okta replied:
This is excellent feedback, thanks @KevinGhadyani-Okta! I don't think I have the chops to add the logic hook myself, so I'll leave that to you or Shred. And again, I just want to be helpful, so if it's counterproductive to have me touching code myself please let me know and I'm happy to leave it to those who actually know what they're doing — I don't want to unintentionally make _more_ work for you by pretending to be an engineer 😅

edburyenegren-okta replied:
MUI exposes `onClose` and accepts any function; if a function is passed, it displays the UI appropriate for dismissal. I think that aligns with your suggestion above without us having to solve for _how_ things are dismissed in various apps.

If that's the case, our stories will need some example code to demo the functionality, but the component itself could ship without an opinion.

<img width="826" alt="Screen Shot 2023-01-12 at 10 48 39 AM" src="https://user-images.githubusercontent.com/36284167/212153295-e19b2b72-5f2c-428c-bf71-c04f9a6a061f.png">

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/Banner/index.ts:
```diff
@@ -0,0 +1,13 @@
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+export * from "./Banner";
```
We don't need these components in separate directories. Right now, I prefer a flat-filesystem which means all files are in `src/`. Theme files are separated out because they technically belong in their own library. All other components should be in `src/`, so we can safely remove this file.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/index.ts:
```diff
@@ -11,3 +11,4 @@
  */
 
 export * from "./TextField";
+export * from "./Banner";
```
Same here, this folder nor `index` shouldn't exist.

KevinGhadyani-Okta replied:
You might wanna pull from `develop` to get the latest code.

KevinGhadyani-Okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Banner/Banner.stories.tsx:
```diff
@@ -79,21 +88,16 @@ Warning.args = {
 };
 
 export const Dismissible = DefaultTemplate.bind({});
 Dismissible.args = {
-  onClose: `{() => {}}`,
+  isDismissable: true,
   severity: "warning",
   role: "status",
 };
```
You already removed `onClose`, but so you know, Storybook adds `onX` functions for you and logs them under Actions for you:
<img width="488" alt="image" src="https://user-images.githubusercontent.com/97464104/212149731-49fcf86a-9ff5-40ab-a742-e3a02a1b981c.png">

All though in our Storybook, this may be disabled; I'm not sure. It's not working from my testing. It may need a ticket to get it added back as that's the default behavior.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/components/Banner/Banner.tsx:
```diff
@@ -0,0 +1,79 @@
+
+import { AlertColor } from "@mui/material";
+import React, { useState } from "react";
+import { Alert, Link } from "../..";
```
I don't think we use `Link` in this component.

jordankoschei-okta commented:
I fixed those issues we discussed, but I'm not sure why CI is failing — can you take a look, @edburyenegren-okta or @KevinGhadyani-Okta ?

edburyenegren-okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,65 @@
+ */
+
+import { AlertColor } from "@mui/material";
+import React, { SyntheticEvent } from "react";
```
This import can be removed now

edburyenegren-okta replied:
You can remove the `React` import from here since we only need `SyntheticEvent`.

edburyenegren-okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,65 @@
+   * The function that's fired when the user clicks the close button. If undefined,
+   * the close button will not be shown.
+   */
+  onClose?: (event: SyntheticEvent<HTMLButtonElement>) => void;
```
This is one of the complaints from CI. Might be an overly specific typing? 

@KevinGhadyani-Okta what's appropriate here?

KevinGhadyani-Okta replied:
There's a mouse event handler type we could use. I have to look it up to be sure, but there are actual types you can use rather than defining the event yourself.

The type itself looks fine, but we're also missing the `() => void` type too. The handler type does that for you.

edburyenegren-okta commented on packages/odyssey-storybook/src/components/odyssey-mui/Banner/Banner.stories.tsx:
```diff
@@ -11,7 +11,7 @@
 
 import { Story } from "@storybook/react";
-import { Alert, Link } from "@okta/odyssey-react-mui";
+import { Alert, Link, Banner } from "@okta/odyssey-react-mui";
```
You can remove `Alert` and `Link`, since we aren't using them here. That'll solve some of the CI complaints.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,64 @@
+   * The function that's fired when the user clicks the close button. If undefined,
+   * the close button will not be shown.
+   */
+  onClose?: any;
```
This should be something like `MouseEventHandler` from React.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,64 @@
+   * ("status" for something that dynamically updates, null for something
+   * unchanging)
+   */
+  role?: "status" | undefined;
```
I figured out what I want here. We should use `AlertProps["role"]` since its' a pass-through props.

If it's a prop we don't control that we simply pass-through, it makes more sense for the consuming component to define the type, not our component.

There's a problem with this approach:
If we start adding functionality in our parent component for this prop, the type is now tied to a child component rather than our own unless someone knows to change it.

But in general, it's a good pattern to avoid tight coupling. If MUI changes their prop, and we have to change ours, that's tight coupling. Ideally, we would set our type to `unknown` or `any` and let MUI handle it, but TypeScript doesn't work that way like runtime PropTypes did.

So in this case, it's best if we let MUI handle the types of any pass-through props like `AlertProps["role"]` rather than re-define those ourselves.

edburyenegren-okta replied:
MUI's type here is `string`, but we don't want people passing in *any* role for Banner, only `status` if necessary. Should we still hinge on MUI's props in that case?

KevinGhadyani-Okta replied:
I guess, why are we passing status or not? Why isn't that hardcoded for `Banner`?

jordankoschei-okta replied:
This is from a conversation Shred and I had last week:
<img width="697" alt="image" src="https://user-images.githubusercontent.com/91091570/213807108-162e4bc3-1cc8-44ba-873a-1c29930b5dfd.png">

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,64 @@
+   * unchanging)
+   */
+  role?: "status" | undefined;
+  /**
+   * The text content of the alert
+   */
+  children: string;
```
`children` should technically never be a string. People will assume you can pass React components in here since just about every `children` is type `ReactNode`. That's how HTML works in general.

I'd recommend using a `text` or `description` prop instead for string-only children just to avoid breaking convention.

Please give me your thoughts. We can meet on this one.

jordankoschei-okta replied:
That makes sense to me, let's switch to `text`.

KevinGhadyani-Okta reviewed (commented):
A few small type things I didn't notice last time; otherwise, we're good.

jordankoschei-okta commented:
@KevinGhadyani-Okta I changed the type of `onClose` and now I'm getting this error. Do you know why?
<img width="1284" alt="image" src="https://user-images.githubusercontent.com/91091570/213807903-6b35791e-73fd-4252-bd8e-90666faa5adb.png">

KevinGhadyani-Okta commented:
That strange. Maybe they're defining the type differently in MUI.

Another option is to set our prop to theirs. `AlertProps["onClick"]`. That might fix it.

Also, you might wanna install this plugin in VSCode:
<img width="575" alt="image" src="https://user-images.githubusercontent.com/97464104/213810666-dbb1b40e-5126-422e-8ec7-9e58bab05b26.png">

It gives you human-readable error messages. Not always helpful, but it gives a second opinion.

KevinGhadyani-Okta commented on packages/odyssey-react-mui/src/Banner.tsx:
```diff
@@ -0,0 +1,65 @@
+   * The text content of the alert
+   */
+  text: string;
+  /**
+   * The function that's fired when the user clicks the close button. If undefined,
+   * the close button will not be shown.
+   */
+  onClose?: ((event: SyntheticEvent<Element, Event>) => void) | undefined;
```
Did we try pulling this from `Alert` since we don't control the type? I don't like using `SyntheticEvent` directly.

KevinGhadyani-Okta replied:
4c2052f7d3f449dafdac892b2e1bbeb8cb7ceacc
